### PR TITLE
Add base_url capability to tljh-config

### LINKED ISSUE #216: Allow running TLJH under a URL prefix
Currently TLJH runs in the root of the domain (example.com), and can not easily work under a URL prefix (example.com/hub). Would be nice to support this.

Requested by @mdogy

### LINKED ISSUE #426: Is possible to make the-littlest-jupyterhub run in other port/url ?
Hi,
I've tried to find the answer in the docs, but with no success.
Is possible to configure the installer to make Jupyter host run in other port (except the 80) or url (i.e. /jupyter) ?

### LINKED ISSUE #616: tljh-config reload should use base url
Hey guys,

https://github.com/jupyterhub/the-littlest-jupyterhub/blob/f0b8745194160d003072f579140dff75358da22d/tljh/config.py#L241

This method makes a request to `http://127.0.0.1:%d/hub/api` and therefore does not work when the `base_url` is changed using `c.JupyterHub.base_url`. I guess this is why `tljh-config reload` hangs after changing the base url.

### PR DESCRIPTION
**Pull request by jeanmarcalkazzi**
This pull request:
* fixes #616 
* fixes #216
* fixes #426

A function update_base_url now sets c.JupyterHub.base_url to the requested value, and correctly checks for it during the check_hub_ready function. 

- [X] Add / update documentation
- [X] Add tests

### CODE REVIEW

**Comment by jeanmarcalkazzi** [comment]
@GeorgianaElena here's the fix for @ikhoury 's issue #616 
Some tests are failing here due to broken master.

**Review by GeorgianaElena — COMMENTED** [review]
Thanks for working on this, @jeanmarcalkazzi, it looks great  :tada: 
What do you think about adding a small integration test in `test_hub.py` to check if a user can login using the new hub url and can start its server?

*Update*: Don't worry about the failing upgrade test, I don't think it's related to this PR. There is a memory limit issue that keeps coming back.

**Comment by jeanmarcalkazzi** [comment]
I will do so in the next few days!
Any specific requirement/aspect to keep in mind for testing?
@GeorgianaElena

**Comment by GeorgianaElena** [comment]
The test should be similar with what happens along these lines: https://github.com/jupyterhub/the-littlest-jupyterhub/blob/master/integration-tests/test_hub.py#L27-L41, setting in addition the base_url.
One thing to keep in mind, all the tests in `test_hub.py` run in the same container, so I'd suggest `unsetting` the new `base_url` after the user server started, so that the other tests won't fail.

**Comment by jeanmarcalkazzi** [comment]
CircleCI output:
```
WARNING: Your kernel does not support swap limit capabilities or the cgroup is not mounted. Memory limited without swap.
b84998b985cfd0e8f45141c9dd0a599ec54c6e313430295724a722be44e0d0de
^@^@Checking if TLJH is already installed...
Setting up hub environment
Installed python & virtual environment
Set up hub virtual environment
Setting up TLJH installer...
Upgraded pip
Setup tljh package
Starting TLJH installer...
Granting passwordless sudo to JupyterHub admins...
Setting up user environment...
Downloading & setting up user environment...
Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/opt/tljh/hub/lib/python3.6/site-packages/tljh/installer.py", line 530, in <module>
    main()
  File "/opt/tljh/hub/lib/python3.6/site-packages/tljh/installer.py", line 502, in main
    ensure_user_environment(args.user_requirements_txt_url)
  File "/opt/tljh/hub/lib/python3.6/site-packages/tljh/installer.py", line 278, in ensure_user_environment
    'conda==' + conda_version
  File "/opt/tljh/hub/lib/python3.6/site-packages/tljh/conda.py", line 108, in ensure_conda_packages
    ] + packages).decode()
  File "/usr/lib/python3.6/subprocess.py", line 356, in check_output
    **kwargs).stdout
  File "/usr/lib/python3.6/subprocess.py", line 438, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['/opt/tljh/user/bin/python', '-m', 'conda', 'install', '-c', 'conda-forge', '--json', '--prefix', '/opt/tljh/user', 'conda==4.8.1']' died with <Signals.SIGKILL: 9>.
Traceback (most recent call last):
  File ".circleci/integration-test.py", line 189, in <module>
    main()
  File ".circleci/integration-test.py", line 173, in main
    run_test(image_name, args.test_name, args.bootstrap_pip_spec, args.test_files, args.upgrade, args.installer_args)
  File ".circleci/integration-test.py", line 108, in run_test
    f'python3 /srv/src/bootstrap.py {installer_args}'
  File ".circleci/integration-test.py", line 71, in run_container_command
    ], check=True)
  File "/usr/lib/python3.6/subprocess.py", line 438, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['docker', 'exec', '-it', 'basic-tests', '/bin/bash', '-c', 'python3 /srv/src/bootstrap.py ']' returned non-zero exit status 1.

Exited with code exit status 1
CircleCI received exit code 1
```

@GeorgianaElena  I think this is related to the memory issue you mentioned.

File: integration-tests/test_hub.py [hunk 1]
```diff
@@ -37,13 +38,41 @@ async def test_user_code_execute():
...
+    assert 0 == await (await asyncio.create_subprocess_exec(*TLJH_CONFIG_PATH, 'reload')).wait()
+
+    async with User(username, hub_url, partial(login_dummy, password='')) as u:
+        await u.login()
+        await u.ensure_server_simulate()
+        await u.start_kernel()
+        await u.assert_code_output("5 * 4", "20", 5, 5)
+
+        # Assert that the user exists
+        assert pwd.getpwnam(f'jupyter-{username}') is not None
```
Comment (GeorgianaElena) [inline integration-tests/test_hub.py:71]: I believe it's enough just to check that the user's server started and we are redirected correctly to it.
```suggestion
```

File: integration-tests/test_hub.py [hunk 2]
```diff
@@ -37,13 +38,41 @@ async def test_user_code_execute():
...
+
+
+@pytest.mark.asyncio
+async def test_user_code_execute_with_custom_base_url():
+    """
+    User logs in, starts a server with a custom base_url & executes code
+    """
+    # This *must* be localhost, not an IP
+    # aiohttp throws away cookies if we are connecting to an IP!
+    hub_url = 'http://localhost'
```
Comment (GeorgianaElena) [inline integration-tests/test_hub.py:57]: The hub url should use the `base_url` we're setting for this test:
```python
hub_url = f"http://localhost{base_url}"
```

**Comment by GeorgianaElena** [comment]
I believe you're right @jeanmarcalkazzi :confused: I will investigate this tomorrow. Thank you!

File: integration-tests/test_hub.py (hunk 2 above)
Comment (jeanmarcalkazzi) [reply integration-tests/test_hub.py]: My bad, now it's added.

**Comment by jeanmarcalkazzi** [comment]
Changes done.
@GeorgianaElena let me know if I can help with the debugging of this memory issue!

**Review by GeorgianaElena — APPROVED** [review]
Hey @jeanmarcalkazzi :wave: The tests were fixed, so I've rebased the PR and now all the bots are happy and everything is green :partying_face: 

I'll go ahead and merge this now. Thanks a lot for your work and patience :sunflower:

**Pull request merged**